sonarts-core/src/se/SymbolicExecution.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * SonarTS
+ * Copyright (C) 2017-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as ts from "typescript";
+import { ControlFlowGraph, CfgBlock, CfgBranchingBlock } from "../cfg/cfg";
+import { applyExecutors } from "./stateTransitions";
+import { ProgramState } from "./programStates";
+import { is, CONDITIONAL_STATEMENTS, LOOP_STATEMENTS } from "../utils/navigation";
+import { getTruthyConstraint, getFalsyConstraint } from "./constraints";
+
+const BLOCK_VISITS_LIMIT = 1000;
+
+export function execute(
+  cfg: ControlFlowGraph,
+  program: ts.Program,
+  initialState: ProgramState,
+): ExecutionResult | undefined {
+  const programNodes = new Map<ts.Node, ProgramState[]>();
+  const branchingProgramNodes = new Map<ts.Node, ProgramState[]>();
+  let visits = 0;
+  visitBlock(cfg.start, initialState);
+  if (visitsLimitBreached()) {
+    // Analysis incomplete, it's safer to not report partial results
+    return;
+  } else {
+    return { programNodes, branchingProgramNodes, visits };
+  }
+
+  function visitBlock(block: CfgBlock, programState: ProgramState) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') The function visitBlock has a cyclomatic complexity of 15 which is higher than the threshold of 10 [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)